label_studio/io_storages/proxy_api.py

```diff
@@ -0,0 +1,327 @@
+            project = instance if isinstance(instance, Project) else instance.project
+            storage_objects = project.get_all_import_storage_objects
+            storage = get_storage_by_url(fileuri, storage_objects)
+            if not storage:
+                logger.error(f'Could not find storage for URI {fileuri}')
+                return Response(status=status.HTTP_404_NOT_FOUND)
+            # Not all storages support presigned URLs
+            if not hasattr(storage, 'presign'):
+                logger.error(f'Storage {storage} does not support presign URLs')
+                return Response(status=status.HTTP_404_NOT_FOUND)
+            presign = storage.presign
+        else:
+            presign = True
+
+        # Check if storage should use presigned URLs;
+        # It's important to have this check here, because it increases security:
+        # If storage.presign is False, it means an admin doesn't want to expose presigned URLs anyhow,
+        # and all files are proxied through Label Studio using LS auth and RBAC control.
+
+        if presign:
+            # Redirect to presigned URL (original flow)
+            return self.redirect_to_presign_url(fileuri, instance, model_name)
+        else:
+            return self.proxy_data_from_storage(request, fileuri, project, storage)
+
+    def redirect_to_presign_url(self, fileuri: str, instance: Union[Task, Project], model_name: str) -> Response:
+        """Generate and redirect to a presigned URL for the given file URI"""
+        try:
+            resolved = instance.resolve_storage_uri(fileuri)
+        except Exception as exc:
+            logger.error(f'Failed to resolve storage uri {fileuri} for {model_name} {instance.id}: {exc}')
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        if resolved is None or resolved.get('url') is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        url = resolved['url']
+        max_age = 0
+        if resolved.get('presign_ttl'):
+            max_age = resolved.get('presign_ttl') * 60
+
+        # Proxy to presigned url
+        response = HttpResponseRedirect(redirect_to=url, status=status.HTTP_303_SEE_OTHER)
+        response.headers['Cache-Control'] = f'no-store, max-age={max_age}'
+        return response
+
+    def time_limited_chunker(self, stream_body):
+        """
+        Generator that stops yielding chunks after timeout seconds without blocking.
+        """
+        chunk_size = settings.RESOLVER_PROXY_BUFFER_SIZE
+        timeout = settings.RESOLVER_PROXY_TIMEOUT
+        
+        start_time = time.monotonic()
+        deadline = start_time + timeout
+        chunks_yielded = 0
+        total_bytes = 0
+        
+        # Flag to indicate if timeout occurred
+        timed_out = False
+        
+        try:
+            for chunk in stream_body.iter_chunks(chunk_size=chunk_size):
+                print(f"==> chunk: {chunks_yielded}, {total_bytes}\n")
```

Comment:
If this is intended for the final version, better to use `logger.info` than `print`